### PR TITLE
fix: align extension/desktop HostInteractions resolve/cancel semantics

### PR DESCRIPTION
**Pull request by LionSR**
## Summary

Fixes the concrete drift points from the 2026-07-07 audit between `packages/extension/src/progressView/extensionHostInteractions.ts` and `packages/desktop/src/main/desktopHostInteractions.ts`:

- **Kind check on resolve**: desktop now rejects a resolution whose `result.kind` does not match the pending request kind, matching the extension host's discriminant check.
- **Shared result vocabulary**: both hosts now use `src/agent/runtime/hostInteractionResultMappers.ts` for bash, plan, proposal, retry, and user-question results instead of maintaining duplicate inline mappers.
- **Proposal pass-through is explicit**: a `result.value` proposal object is trusted only for declared `ProposalResult` actions, including `timeout`; arbitrary `{ action: ... }` objects fall back to the resolution action and feedback.
- **Cancellation cause forwarding**: extension cancellation now routes through the same mappers as live resolution, forwarding the cause for bash, plan, proposal, and user-question results. Retry cancellation still resolves to `{ action: 'cancel' }`, because `RetryResult` has no feedback field on that variant.
- **Bash timeout preservation**: after #7492, the shared bash mapper preserves `{ action: 'timeout' }` as `timedOut: true`, keeping timeout distinct from explicit rejection.

The shared mapper deliberately follows the stricter extension semantics for bash and user-question fields: bash `userMessage` is attached to rejection only, feedback is trimmed, submit questions carry answers, and non-submit questions carry feedback. That keeps both hosts on one interpretation rather than preserving the older desktop-only permissive pass-through.

Scope is deliberately bounded: this PR does not change either host's pending-request bookkeeping map.

## Verification

- `corepack pnpm exec eslint src/agent/runtime/hostInteractionResultMappers.ts packages/extension/src/progressView/extensionHostInteractions.ts packages/desktop/src/main/desktopHostInteractions.ts src/test-kernel/progressView/ExtensionHostInteractions.vitest.mts src/test-kernel/desktop/DesktopHostInteractions.vitest.mts && git diff --check`
- `CI=true corepack pnpm exec vitest run src/test-kernel/progressView/ExtensionHostInteractions.vitest.mts src/test-kernel/desktop/DesktopHostInteractions.vitest.mts`
- `CI=true corepack pnpm exec vitest run src/test-kernel/desktop/DesktopAgentExecution.vitest.mts -t "routes runtime events to the window-local desktop backend" --testTimeout=30000`
- `CI=true corepack pnpm exec tsc --noEmit --pretty false`
- `CI=true corepack pnpm exec tsc --noEmit -p tsconfig.test-kernel.json --pretty false`

Addresses #7447.

---

> [!NOTE]
> **Medium Risk**
> Changes agent-visible approval, rejection, timeout, and cancellation outcomes for both desktop and extension hosts; behavior is intentionally tightened in a few desktop edge cases, though scope stays in host interaction mapping with dedicated tests.
> 
> **Overview**
> Unifies how the **desktop** and **extension** `HostInteractions` implementations turn generic UI resolutions and cancellations into typed agent results, closing audit drift between the two hosts.
> 
> A new **`hostInteractionResultMappers`** module is the single place that maps `HostInteractionResolution` into bash, plan, proposal, retry, and user-question outcomes. Both hosts call these mappers for live **`resolve()`** and for synthesized rejections on **cancel/dispose**, so cancellation **`cause`** text reaches the agent the same way a UI rejection would (e.g. bash **`userMessage`**, plan/proposal **`feedback`**).
> 
> **Desktop** now rejects **`resolve()`** when **`result.kind`** does not match the pending request (matching extension), routes **`dispose()`** through **`cancelAll`** with a stable message instead of a special dispose resolution, and adopts the stricter shared rules for bash (**`userMessage`** only on reject; **`timeout`** → **`timedOut`**) and user questions (answers vs feedback tied to submit vs decline). **Extension** cancellation no longer ignores **`cause`**; it uses the same mappers as **`resolve()`**.
> 
> **Proposal** **`result.value`** pass-through is guarded: only objects whose **`action`** is a known **`ProposalResult`** action (including **`timeout`**) are trusted; otherwise mapping falls back to **`result.action`** / **`feedback`**.
> 
> Vitest coverage is added/extended for kind mismatches, cancellation cause forwarding, bash timeout, proposal whitelist behavior, and dispose.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 76a6c450a1e1286b5fba4ee3a96f4d6a05a43985. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by LionSR** [comment]
Updated after review and #7492 rebase: preserved bash timeout as `timedOut: true`, made the proposal action guard type-honest by accepting the declared `timeout` action, imported the real `HostInteractionResolution` type in the desktop test, and rewrote the PR body to document the intentional extension-aligned mapper semantics plus the retry-cancel feedback limitation.

**Comment by LionSR** [comment]
All 5 findings checked. Status:

1. **Object.hasOwn + typeof guard** (copilot) — real, fixed in 76a6c450a: `isProposalResultValue` now uses `Object.hasOwn(value, 'action')` + an explicit `typeof action === 'string'` check instead of `'action' in value` + a bare cast, matching the `toolKind.ts` convention.
2 & 3. **Bash userMessage / userQuestion answers-feedback gating "behavioral change"** (texra-review) — deliberate, not accidental: reviewed both fields' actual purpose (userMessage explains a *rejection*; answers only make sense once *submitted*, feedback explains a *decline*) and confirmed the old desktop truthiness-based gating was itself the drift, not a feature worth preserving. Documented the reasoning directly in the mapper (76a6c450a) rather than relaxing behavior.
4. **PR description overclaim on retry-cancel cause** (texra-review) — already accurate: the description already states "Retry cancellation still resolves to `{ action: 'cancel' }`, because `RetryResult` has no feedback field on that variant" — no change needed.
5. **ProposalResult predicate excluding 'timeout'** (texra-review) — already fixed by an earlier commit on this branch (6a98e03a0), which added `'timeout'` to `PROPOSAL_RESULT_ACTIONS`. Verified against the current file content.

Verification: `npm run typecheck` clean, `eslint` clean, `vitest run` on ExtensionHostInteractions + DesktopHostInteractions — 15/15 passing.

File: src/agent/runtime/hostInteractionResultMappers.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import type { RetryResult } from './RetryRequestCoordinator';
+
+export function toBashApprovalResult(
+  result: HostInteractionResolution,
+): HostBashApprovalResult {
+  return {
+    accepted: result.action === 'approve',
+    timedOut: result.action === 'timeout' ? true : undefined,
+    // Deliberate alignment choice, not an oversight: `userMessage` exists to
+    // explain a rejection back to the agent (see buildBashApprovalRejectedResult),
```
[the remark this answers is not in the document]
Comment (LionSR) [reply src/agent/runtime/hostInteractionResultMappers.ts:30]: buildBashApprovalRejectedResult does exist — src/tools/approval/bashApproval.ts:113, added by #7492 (merged after this review ran, before this branch rebased onto it). Verified via a repo-wide grep against fresh origin/main. Stale-context false positive, no change needed.
